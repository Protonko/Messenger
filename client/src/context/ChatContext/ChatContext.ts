import {createContext} from 'react'

export interface IChatContext {
  selectedMessagesIds: string[]
  toggleSelectMessageId: (selectedMessageId: string) => void
  resetSelectedMessagesIds: () => void
}

export const defaultChatContextValue: IChatContext = {
  selectedMessagesIds: [],
  toggleSelectMessageId: () => {},
  resetSelectedMessagesIds: () => {},
}

export const ChatContext = createContext<IChatContext>(defaultChatContextValue)
