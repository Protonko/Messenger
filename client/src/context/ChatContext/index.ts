export * from './ChatContext'
export * from './ChatContextProvider'
