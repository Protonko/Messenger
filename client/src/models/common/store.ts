export interface IAction<T extends string, U = undefined, V = undefined> {
  type: T
  payload: U
  meta: V
}

export interface IBuildReducer<T, U, V extends string, M = undefined> {
  (state: T, action: IAction<V, U, M>): T
}

export type TReducers<T, U, V extends string, M = undefined> = Record<string, IBuildReducer<T, U, V, M>>
