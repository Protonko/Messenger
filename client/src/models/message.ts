import type {TStatusResponse} from './common/fetch'

export interface ICreateMessageBody {
  id: string
  text: string
}

export interface ICreateMessageResponse {
  status: TStatusResponse
  read: boolean
  // TODO: FIX
  attachments: Array<any>
  id: string
  user: string
  createdAt: string
  updatedAt: string
}

export interface IMessage {
  read: boolean
  // TODO: FIX
  attachments: Array<any>
  id: string
  user: string
  createdAt: string
  updatedAt: string
}
