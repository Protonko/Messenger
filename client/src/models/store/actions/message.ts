import type {AnyAction} from 'redux'
import type {IMessage} from 'models/message'

export enum MessageActionsTypes {
  CREATE_MESSAGE = '[MESSAGE]CREATE_MESSAGE',
  CREATE_MESSAGE_SUCCESS = '[MESSAGE]CREATE_MESSAGE_SUCCESS',
  CREATE_MESSAGE_ERROR = '[MESSAGE]CREATE_MESSAGE_ERROR',
  GET_MESSAGES = '[MESSAGE]GET_MESSAGES',
  GET_MESSAGES_SUCCESS = '[MESSAGE]GET_MESSAGES_SUCCESS',
  GET_MESSAGES_ERROR = '[MESSAGE]GET_MESSAGES_ERROR',
  APPEND_MESSAGE = '[MESSAGE]APPEND_MESSAGE',
}

// payloads
export interface IGetMessagesSuccessPayload {
  messages: IMessage[]
  dialogId: string
}

export interface IGetMessagesErrorPayload {
  errorMessage: string
  dialogId: string
}

export interface ICreateMessagePayload {
  text: string
  dialogId: string
  interlocutorId: string
}

// actions
export interface CreateMessageAction extends AnyAction {
  type: MessageActionsTypes.CREATE_MESSAGE
  payload: ICreateMessagePayload
}

export interface CreateMessageSuccessAction extends AnyAction {
  type: MessageActionsTypes.CREATE_MESSAGE_SUCCESS
  payload: IMessage
}

export interface CreateMessageErrorAction extends AnyAction {
  type: MessageActionsTypes.CREATE_MESSAGE_ERROR
  payload: string
}

export interface GetMessagesAction extends AnyAction {
  type: MessageActionsTypes.GET_MESSAGES
  payload: string
}

export interface GetMessagesSuccessAction extends AnyAction {
  type: MessageActionsTypes.GET_MESSAGES_SUCCESS
  payload: IGetMessagesSuccessPayload
}

export interface GetMessagesErrorAction extends AnyAction {
  type: MessageActionsTypes.GET_MESSAGES_ERROR
  payload: IGetMessagesErrorPayload
}

export interface AppendMessageAction extends AnyAction {
  type: MessageActionsTypes.APPEND_MESSAGE
  payload: IMessage
}

export type AllMessageActions =
  | CreateMessageAction
  | CreateMessageSuccessAction
  | CreateMessageErrorAction
  | GetMessagesAction
  | GetMessagesSuccessAction
  | GetMessagesErrorAction
  | AppendMessageAction
