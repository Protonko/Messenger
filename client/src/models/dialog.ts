import type {IUser} from './user'
import {Status, ReadStatus} from './common/status'

export interface IDialog {
  id: string
  interlocutor: IUser,
  lastMessage: string
  createdAt: string
  updatedAt: string
  messages: number
  status: Status
  readStatus: ReadStatus | null
}

export interface ICreateDialogBody {
  author: string
  interlocutor: string
  text: string
}
