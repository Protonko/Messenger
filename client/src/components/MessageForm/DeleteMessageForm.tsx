import {useContext} from 'react'
import {useDispatch} from 'react-redux'
import {Button, ButtonModifier} from 'components/common/Button'
import {ChatContext} from 'context/ChatContext'
import {deleteMessages} from 'store/actions/message'
import {useSearchParams} from 'hooks/useSearchParams'

export const DeleteMessageForm = () => {
  const dispatch = useDispatch()
  const dialogId = useSearchParams('dialog')
  const {resetSelectedMessagesIds, selectedMessagesIds} =
    useContext(ChatContext)

  const onDelete = () => {
    if (dialogId) {
      dispatch(deleteMessages({messagesIds: selectedMessagesIds, dialogId}))
      resetSelectedMessagesIds()
    }
  }

  return (
    <div className="delete-message-form">
      <Button
        additionalClassName="delete-message-form__button delete-message-form__button--delete"
        text="DELETE"
        onClick={onDelete}
      />
      <Button
        additionalClassName="delete-message-form__button delete-message-form__button--cancel"
        modifier={ButtonModifier.LINKING}
        text="CANCEL"
        onClick={resetSelectedMessagesIds}
      />
    </div>
  )
}
