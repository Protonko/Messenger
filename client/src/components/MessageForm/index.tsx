import {useContext} from 'react'
import {ChatContext} from 'context/ChatContext'
import {DeleteMessageForm} from './DeleteMessageForm'
import {CreateMessageForm} from './CreateMessageForm'

export const MessageForm = () => {
  const {selectedMessagesIds} = useContext(ChatContext)

  if (selectedMessagesIds.length) {
    return <DeleteMessageForm />
  }

  return <CreateMessageForm />
}
