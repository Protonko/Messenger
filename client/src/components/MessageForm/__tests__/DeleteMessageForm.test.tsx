import type {MockStoreEnhanced} from 'redux-mock-store'
import type {RootState} from 'store/reducers'
import {mount, ReactWrapper} from 'enzyme'
import {Provider} from 'react-redux'
import {Router} from 'react-router-dom'
import {createMemoryHistory, History, Location} from 'history'
import {MessageActionsTypes} from 'models/store/actions/message'
import {DeleteMessageForm} from 'components/MessageForm/DeleteMessageForm'
import {ChatContext, IChatContext} from 'context/ChatContext'
import {DIALOG} from 'static/test-mocks/dialog'
import {createMockStore, state} from 'static/test-mocks/store'

const ELEMENT_SELECTORS = {
  deleteButton: 'delete-message-form__button--delete',
}

describe('DeleteMessageForm', () => {
  let component: (store: MockStoreEnhanced<RootState>) => ReactWrapper
  let history: History

  const historyLocation: Location = {
    pathname: '',
    search: `dialog=${DIALOG.id}`,
    hash: '',
    key: '',
    state: null,
  }

  const contextValue: IChatContext = {
    selectedMessagesIds: ['123'],
    toggleSelectMessageId: jest.fn(),
    resetSelectedMessagesIds: jest.fn(),
  }

  beforeEach(() => {
    history = {
      ...createMemoryHistory(),
      push: jest.fn(),
      goBack: jest.fn(),
      go: jest.fn(),
      location: historyLocation,
    }

    component = (store) => {
      return mount(
        <Provider store={store}>
          <Router history={history}>
            <ChatContext.Provider value={contextValue}>
              <DeleteMessageForm />
            </ChatContext.Provider>
          </Router>
        </Provider>,
      )
    }
  })

  it('Should render component', () => {
    expect(component(createMockStore(state))).toMatchSnapshot(
      'DeleteMessageForm',
    )
  })

  it('Should call dispatch on delete button click', () => {
    const payload = {
      messagesIds: ['123'],
      dialogId: 'id',
    }

    const store = createMockStore(state)
    store.dispatch = jest.fn()
    const componentMount = component(store)
    componentMount.find(`.${ELEMENT_SELECTORS.deleteButton}`).simulate('click')

    expect(store.dispatch).toBeCalledWith({
      type: MessageActionsTypes.DELETE_MESSAGES,
      payload,
    })
  })
})
