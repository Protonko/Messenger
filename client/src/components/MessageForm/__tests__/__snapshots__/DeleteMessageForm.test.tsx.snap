// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteMessageForm Should render component: DeleteMessageForm 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "i",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [MockFunction],
        "goBack": [MockFunction],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "",
          "pathname": "",
          "search": "dialog=id",
          "state": null,
        },
        "push": [MockFunction],
        "replace": [Function],
      }
    }
  >
    <DeleteMessageForm>
      <div
        className="delete-message-form"
      >
        <Button
          additionalClassName="delete-message-form__button delete-message-form__button--delete"
          onClick={[Function]}
          text="DELETE"
        >
          <button
            className="button delete-message-form__button delete-message-form__button--delete"
            onClick={[Function]}
            type="button"
          >
            <span
              className="button__text"
            >
              DELETE
            </span>
          </button>
        </Button>
        <Button
          additionalClassName="delete-message-form__button delete-message-form__button--cancel"
          modifier="linking"
          onClick={[MockFunction]}
          text="CANCEL"
        >
          <button
            className="button button--linking delete-message-form__button delete-message-form__button--cancel"
            onClick={[MockFunction]}
            type="button"
          >
            <span
              className="button__text"
            >
              CANCEL
            </span>
          </button>
        </Button>
      </div>
    </DeleteMessageForm>
  </Router>
</Provider>
`;
